Move pulumi context settings and persistent data

JimMadge opened the pull request:
## :white_check_mark: Checklist

<!--
Thank you for your pull request!

Before going any further, please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.
Replace the empty checkboxes [ ] below with checked ones [x] accordingly.
-->

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).
- [x] You have marked this pull request as a **draft** and added `'[WIP]'` to the title if needed (if you're not yet ready to merge).

### :arrow_heading_up: Summary

<!--
Please explain what your pull request does here.
You might (optionally) want to include screenshots here.
-->

- Add `ConfigClass` base class for configuration synced to Azure Storage in the Context
  - Add `from_remote_or_create` method to fetch config if it exists in Azure Storage, otherwise create a new instance
- Add `DSHPulumiProject` and `DSHPulumiConfig` classes for storing persistent Pulumi project information
- Use `ConfigClass` for both `Config` and `DSHPulumiConfig`
- Remove the need for `Context` to construct `Config`. Removes a lot of potential circular imports.
- Use information from `DSHPulumiProject` instances to create Pulumi workspaces and stacks, avoiding the need to write `Pulumi.<stack>.yaml` and `Pulumi.yaml` files in a non-temp work directory
- Propagated above changes through infrastructure and commands modules
- Many tests
  - Adding initial tests for `StackManager` classes
  - Add mocking frameworks for testing some Azure/Pulumi functions

### :closed_umbrella: Related issues

<!--
If your pull request will close any open issues (hopefully it will!) then add `Closes #<issue number>` here.
-->

Closes #1689 #1828 

### :microscope: Tests

<!--
- Document any manual tests that you have carried out (*e.g.* deploying a new SHM and/or SRE) and confirm which commit you did this with
- Note that automated tests will be run as part of the CI process and will block this PR until they pass.
- Additionally, a successful code review may be required before this PR can be merged.
- If this pull request only changed documentation you can simply state 'Documentation only' here.
-->

New deployment, repeating commands successfully fetches Pulumi persistent data (stack settings, stack state)

JimMadge committed:
Move Pulumi storage account and key to context

JimMadge committed:
Add test for pulumi_backend_url

JimMadge committed:
Add test for pulumi_secrets_provider_url

JimMadge committed:
Add tests  for context properties

JimMadge committed:
Add basic pulumi configuration class

JimMadge committed:
Run lint:fmt

JimMadge committed:
Add example class test for UniqueList

JimMadge committed:
Add model_dump test

JimMadge committed:
Use UniqueList for pulumi stacks

JimMadge committed:
Add getitem method for PulumiConfig

JimMadge committed:
Run lint:fmt

JimMadge committed:
Add stack_names method

JimMadge committed:
Add to_yaml method

JimMadge committed:
Add delitem method

JimMadge committed:
Add extra delitem tests

JimMadge committed:
Add setitem method

JimMadge committed:
Fix typing

JimMadge committed:
Add eq tests

JimMadge committed:
Add docstrings

JimMadge committed:
Add write_config method

JimMadge committed:
Add from_yaml classmethod

JimMadge committed:
Add upload method

JimMadge committed:
Add from_remote classmethod

JimMadge committed:
Remove Pulumi persistent data from Config class

JimMadge commented:
Working on this, the names are becoming a mess.

Terms [here](https://www.pulumi.com/docs/using-pulumi/automation-api/concepts-terminology/).

We have two Pulumi Programs, one for SHM, one for SRE.
Each 'thing' we deploy is a new Pulumi Project, with a Local Workspace containing project configuration `Pulumi.yaml` and _a single_ stack with configuration `Pulumi.<stack>.yaml`. (We might not need the Local Workspace at all #1828)

We _could_ use stacks a bit more like they are intended, with dev and prod deployments. I don't see us doing that now though. So, the state is, each 'thing' is the only stack of a Pulumi Project.

I think that has lead to conflating the terms program/stack and not using the term project.

I propose

- `DSHPulumiConfig` - Class to hold persistent Pulumi information that is written to the config share (multiple instances of `DSHPulumiProject`. _I.e._ _not_ stacks which are in the Pulumi backend, _not_ the Pulumi backend address/encryption key which are in the context
- `DSHPulumiProject` - Class to hold persistent information about Pulumi projects, which right now is just a name and a single stack configuration (like `Pulumi.<stack>.yaml`)
- We should be more strict and consistent with naming and be careful not to call projects stacks

@craddm @jemrobinson thoughts?

JimMadge commented:
Also, I must be tired because I implemented a class where you can look up items using strings, which must be unique, and didn't use a `dict` 🤦.

jemrobinson commented:
Can you give a quick sketch of what would be in each of these config files? Could be e.g. a YAML representation of the pieces of data in each of them.

JimMadge commented:
> Can you give a quick sketch of what would be in each of these config files?

The Pulumi files, or the classes we are creating?

For the later, I would write the `DSHPulumiConfig` class to the storage account in the context in a similar way to `Config`, taking advantage of the Pydantic serializer.

jemrobinson commented:
> The Pulumi files, or the classes we are creating?

The data structure of the classes we're creating (I'm not worried about the serialisation part). I'm thinking something like:

```yaml
DSHPulumiConfig
- azure
  - subscription_id
  - tenant_id
- something_else
```

```yaml
DSHPulumiProject
- something
  - A
  - B
- something_else
  - C
  - D  
```

JimMadge commented:
Currently,

```
DSHPulumiConfig:
  projects:
    - DSHPulumiProject: # e.g. an SHM or an SRE
      - name: a
      - config: ...
    - DSHPulumiProject: # e.g. an SHM or an SRE
      - name: b
      - config: ...
```

jemrobinson commented:
What do we gain by allowing multiple `DSHPulumiProject` to live under one `DSHPulumiConfig`? As written, there are no keys which belong to the top-level `DSHPulumiConfig` object except instances of `DSHPulumiProject`. I'm sure there's more to what you're suggesting than that.

jemrobinson commented on data_safe_haven/config/pulumi.py:
```diff
@@ -0,0 +1,147 @@
+
+    def __getitem__(self, key: str) -> PulumiStack:
+        if not isinstance(key, str):
+            msg = "'key' must be a string."
```
```suggestion
            msg = "'{key}' must be a string."
```

JimMadge replied:
This is intentional, 'key' the argument rather than `{key}` the value.

JimMadge commented:
> What do we gain by allowing multiple `DSHPulumiProject` to live under one `DSHPulumiConfig`? As written, there are no keys which belong to the top-level `DSHPulumiConfig` object except instances of `DSHPulumiProject`. I'm sure there's more to what you're suggesting than that.

Being able to look up projects by a key, which I believe we need when Pulumi outputs are shared between projects, `dsh_pulumi_config['my_shm']`, `dsh_pulumi_config['my_sre']`. Serialising all Pulumi projects and writing them to a file in the context.

It is essentially a container at the moment, but we want it to have move functionality than just a plain `dict` or `list`. If there is more persistent Pulumi data, default args for example, this would be a natural place to store them.

JimMadge committed:
Revert "Remove Pulumi persistent data from Config class"

This reverts commit f0f97760dd0206f660d1682aafdbed4a00e57688.

JimMadge committed:
Run lint:fmt

JimMadge committed:
Add ConfigClass

JimMadge committed:
Restructure pulumi config to use ConfigClass

JimMadge committed:
Use ConfigClass for Config

JimMadge committed:
Restructure pulumi config classes

JimMadge committed:
Add stack_config_dict method

JimMadge committed:
Remove ConfigSectionPulumi and methods

jemrobinson commented on data_safe_haven/administration/users/guacamole_users.py:
```diff
@@ -3,18 +3,21 @@
 from typing import Any
 
 from data_safe_haven.config import Config
+from data_safe_haven.config.context_settings import Context
```
As a rule-of-thumb (can't find a reference to it at the moment) it's better not to use deeply-nested imports in Python code. This makes your dependencies more fragile and means that it's harder to refactor. In this case, I think we'd be better with the following:

```suggestion
from data_safe_haven.config import Config, Context
```

with an equivalent change to `config > __init__.py`.

JimMadge replied:
I did this because there were circular dependencies.

I think that might not be a problem any more so I'll look to tidy that.

jemrobinson replied:
For local dependencies I prefer `from .context_settings import Context`. If they're really circular, we should think about how to break the coupling.

JimMadge replied:
Done in 2012ff6

jemrobinson commented on data_safe_haven/infrastructure/stack_manager.py:
```diff
@@ -57,20 +59,22 @@ class StackManager:
     ) -> None:
-        self.account = PulumiAccount(config)
+        self.account = PulumiAccount(context, config)
+        self.context: Context = context
         self.cfg: Config = config
```
Not sure you need type markers here since the type is defined in the function signature.

JimMadge committed:
Remove old tests, move fixtures

JimMadge committed:
Move more fixtures

JimMadge committed:
Move sanitite sre name function

JimMadge committed:
Add basic tests for StackManager/SHMStackManager

JimMadge committed:
Restructure DSHPulumiProject class

JimMadge committed:
Update create_or_select_stack

JimMadge committed:
Add stack_all_config property

JimMadge committed:
Add udpate_dsh_pulumi_project method

JimMadge committed:
Add create_or_select_project method

JimMadge committed:
Propagate Pulumi config changes to infra/commands

JimMadge committed:
Merge remote-tracking branch 'origin/develop' into pulumi_storage

JimMadge committed:
Remove unnecessary type hints

JimMadge committed:
WIP rearrange module structure

jemrobinson commented on data_safe_haven/config/config_class.py:
```diff
@@ -0,0 +1,69 @@
+T = TypeVar("T", bound="ConfigClass")
+
+
+class ConfigClass(BaseModel, validate_assignment=True):
```
I prefer more descriptive class names if possible. How about?

```suggestion
class SerialisableConfig(BaseModel, validate_assignment=True):
```

JimMadge replied:
Done in 7a8cc97

jemrobinson commented on data_safe_haven/functions/__init__.py:
```diff
@@ -35,6 +36,7 @@
     "random_letters",
     "replace_separators",
     "seeded_uuid",
+    "sanitise_sre_name",
```
```suggestion
    "sanitise_sre_name",
    "seeded_uuid",
```

jemrobinson reviewed (changes requested):
I haven't gone through this in detail but I've got a couple of general thoughts.

1. I prefer to keep one class per file, with the filename as a snake-case name of the class. This makes it a bit easier (for me) to find where things are defined. Do you think that's sensible?
2. I saw you've replaced the local workdir setting - does this work?
3. I can't get this to work! Have you tested it? If so, can you give me some instructions about how to do it?

JimMadge committed:
Update imports

jemrobinson commented on data_safe_haven/commands/deploy.py:
```diff
@@ -28,6 +31,8 @@ def shm(
     """Deploy a Safe Haven Management component"""
     context = ContextSettings.from_file().assert_context()
     config = Config.from_remote(context)
```
What if there's no remote file (e.g. on first run)?

JimMadge replied:
Fixed in 65ad597.

I've only put it in for deploy SHM for now (as the other commands couldn't work without this file already existing, I think).

jemrobinson commented on data_safe_haven/commands/deploy.py:
```diff
@@ -28,6 +31,8 @@ def shm(
     """Deploy a Safe Haven Management component"""
     context = ContextSettings.from_file().assert_context()
     config = Config.from_remote(context)
+    pulumi_config = DSHPulumiConfig.from_remote(context)
```
What if there's no remote file (e.g. on first run)?

JimMadge replied:
Good point. I'll add something like "get or create DSHPulumiConfig" :+1:.

jemrobinson replied:
What command creates the remote Pulumi config?

JimMadge replied:
To write it to Azure Storage? `DSHPulumiConfig.upload` but you would need to check if one already exists and create an empty instance of the class if not.

jemrobinson replied:
I'm thinking about the workflow that someone would have when trying to deploy this. Something like:

```
dsh context add # makes the local context
dsh context create # deploys Azure resources
dsh config template > config.yaml # creates a local YAML file
dsh config upload config.yaml # upload the local config to Azure
dsh deploy shm
```

this last line breaks because it looks for a remote Pulumi config but none of the previous commands have created it.

Also - this workflow seems a bit overcomplicated, but we can fix that elsewhere.

JimMadge replied:
I see, I would have it happen at `$ dsh deploy shm`.

JimMadge replied:
I prefer to keep the workflow explicit like that. I wouldn't like the command which adds a description of a context to also deploy infrastructure.

jemrobinson replied:
Fair enough. Perhaps we could replace `dsh config upload config.yaml` with something like `dsh deploy shm --config config.yaml` which would prompt for "are you sure?" if there's a remote config that differs from `config.yaml`.

JimMadge replied:
I think that sounds sensible. It would be easy (and probably useful) to keep the `$ dsh config upload ...` command. Both functions should just need to call a couple of `Config` methods.

Could you open an issue, if there isn't one already.

jemrobinson replied:
If this is in a usable state by this afternoon, we can discuss it during a live run-through.

jemrobinson replied:
I'm happy to have more powerful advanced commands (or options to commands) but I'd like to keep the minimal required commands simple and easy to understand. If possible, they would just be: `dsh initialise; dsh deploy` but I realise we're not there at the moment!

JimMadge replied:
I think this is a philosophical argument.

I would say that isn't simple or easy to understand. Having so few commands means they must be complex and handle a lot of processes without input from the user.

I would say something more like a tree (`dsh shm {deploy, destroy}`, `dsh sre {deploy, destroy}` is simple and easy to understand because the commands are clear, limited and descriptive.

Could you open an issue so I can resolve the conversation here?

jemrobinson replied:
See #1836 for discussion.

JimMadge committed:
Update data_safe_haven/functions/__init__.py

Co-authored-by: James Robinson <james.em.robinson@gmail.com>

JimMadge commented:
CI is failing because Pulumi non-interactive login is failing. Not sure if there is a way to do a non-interactive local login.

jemrobinson commented:
> CI is failing because Pulumi non-interactive login is failing. Not sure if there is a way to do a non-interactive local login.

Can we mock it?

JimMadge committed:
Enforce local Pulumi account for tests

An autouse fixture with scope for the entire test session logs in to a
local Pulumi account, then logs out.

JimMadge commented:
> > CI is failing because Pulumi non-interactive login is failing. Not sure if there is a way to do a non-interactive local login.
> 
> Can we mock it?

No, or, not unless we mock the entire Pulumi CLI :smile:.

jemrobinson commented:
> > > CI is failing because Pulumi non-interactive login is failing. Not sure if there is a way to do a non-interactive local login.
> > 
> > 
> > Can we mock it?
> 
> No, or, not unless we mock the entire Pulumi CLI 😄.

This might be time to take a deeper look at the Pulumi testing framework (especially the unit-testing part: https://www.pulumi.com/docs/using-pulumi/testing/unit/).

... in another issue though, preferably.

JimMadge commented:
I think I've fixed that, although there is another problem.
It works locally, so I think _that_ part is OK but it might be failing due to lack of Pulumi/Azure CLI credentials.

JimMadge committed:
Add check keyword arg to subprocess run

JimMadge committed:
Add blob_exists method

JimMadge committed:
Add from_remote_or_create method

JimMadge committed:
Create DSHPulumiConfig on SHM deploy if needed

JimMadge committed:
Correct blob_exists method

JimMadge committed:
Use no secrets provider for some tests

JimMadge committed:
Try None instead of lambda: None

JimMadge committed:
Add CONFIG_PASSPHRASE env var for tests

JimMadge committed:
Rename ConfigClass to SerialisableConfig

JimMadge committed:
Move test file structure consistent with project

JimMadge commented:
> I prefer to keep one class per file, with the filename as a snake-case name of the class. This makes it a bit easier (for me) to find where things are defined. Do you think that's sensible?

I wouldn't be so strict as one class per file. That might end up with lots of very short files. I find it more difficult to find things if I have to start guessing at filenames or using grep. It also means you might need a lot of buffers open to make a small change.

However, I like files to have a clear purpose and not be very long.

craddm committed:
Add clearer error for context show when no context present

craddm committed:
Add wait between creating keyvault and adding key

craddm committed:
add link to list of azure vm SKUs to config template

craddm committed:
Add index field to sre config template

craddm committed:
Ensure upload of Pulumi config when deploying shm

craddm committed:
Adding encrypted key to pulumi yaml

jemrobinson committed:
:sparkles: Add a YAML-serialisable model

jemrobinson committed:
:truck: Move SerialisableConfig to AzureSerialisableModel

jemrobinson committed:
:recycle: Make ContextSettings inherit from YAMLSerialisableModel

jemrobinson committed:
:loud_sound: Improve ContextSettings docstring

jemrobinson committed:
:loud_sound: Add file-level docstring

JimMadge committed:
Run tests and linting on all PRs

JimMadge committed:
Run tests and linting on all PRs

jemrobinson committed:
:bug: Fixed import structure to avoid circular imports

jemrobinson committed:
:truck: Moved tests for SerialisableConfig into tests for AzureSerialisableConfig

jemrobinson committed:
:bug: Fix some outdated test strings

jemrobinson committed:
:bug: Fix YAML dump settings

jemrobinson committed:
:sparkles: Add new test for YAMLSerialisableModel

jemrobinson committed:
:rotating_light: Apply linting changes

jemrobinson committed:
:bug: Added missing config_type for ContextSettings class

jemrobinson committed:
:white_check_mark: Add tests for to_filepath and from_filepath

JimMadge committed:
Use fixture in to_filepath test

jemrobinson committed:
Merge pull request #1840 from alan-turing-institute/separate-config-classes

Separate config classes

JimMadge committed:
Merge branch 'pulumi_storage' into pulumi_storage_fixes

JimMadge committed:
Add test for no context file

JimMadge committed:
Improve message and exit code

JimMadge committed:
Fix DSHPulumiProject fixtures

JimMadge committed:
Update tests

JimMadge committed:
Fix linting

JimMadge committed:
Ensure pulumi configuration is uploaded for sre

JimMadge committed:
Add test ensuring upload is called

JimMadge committed:
Replace unittest.mock with mocker

JimMadge committed:
Run lint:fmt

JimMadge committed:
Merge pull request #1839 from craddm/pulumi_storage_fixes

Pulumi storage fixes, including adding encryption key

JimMadge committed:
Add missing test file

JimMadge committed:
Correct arguments to create_or_select_stack

JimMadge committed:
Add fixture to fix stack manager tests

JimMadge committed:
Add tests for project and stack settings

JimMadge committed:
Run lint:fmt

JimMadge commented:
This is working now and I think I've addressed all the comments.
